builtin/checkout-index.c

```diff
@@ -128,11 +128,18 @@ static int checkout_all(const char *prefix, int prefix_length, int include_spars
 	int i, errs = 0;
 	struct cache_entry *last_ce = NULL;
 
-	if (include_sparse)
-		ensure_full_index(&the_index);
-
 	for (i = 0; i < active_nr ; i++) {
 		struct cache_entry *ce = active_cache[i];
+		if (include_sparse && S_ISSPARSEDIR(ce->ce_mode)) {
+			/*
+			 * If the current entry is a sparse directory (and entries outside the
+			 * sparse checkout definition are included), expand the index and
+			 * continue the loop on the current index position (now pointing to the
+			 * first entry inside the expanded sparse directory).
+			 */
+			ensure_full_index(&the_index);
+			ce = active_cache[i];
```

Comment:
This works because this is the first sparse directory in the index (no earlier entries expand to multiple entries).

This means that we will hit this while iterating over a sparse index, and this commit does not actually reduce the number of expansions, just makes it lazy.

I will look to a later commit to see if this code movement makes an impact there.

Reply:
Now that I see in the following commit that you are able to have a sparse index without sparse directory entries, this makes sense. I would swap the order of the commits, so we can refer to this possible state when making this change.